src/pool/srv_pool.c

```diff
@@ -1081,6 +1084,48 @@ fini_svc_pool(struct pool_svc *svc)
 	svc->ps_pool = NULL;
 }
 
+/*
+ * There might be some swim status inconsistency, let's check and
+ * fix it.
+ */
+static int
+pool_svc_check_node_status(struct pool_svc *svc)
+{
+	struct pool_domain	*doms;
+	int			doms_cnt;
+	int			i;
+	int			rc = 0;
+
+	doms_cnt = pool_map_find_nodes(svc->ps_pool->sp_map, PO_COMP_ID_ALL,
+				       &doms);
+	D_ASSERT(doms_cnt >= 0);
+	for (i = 0; i < doms_cnt; i++) {
+		struct swim_member_state state;
+
+		/* Only check if UPIN server becomes DEAD for now */
+		if (!(doms[i].do_comp.co_status & PO_COMP_ST_UPIN))
+			continue;
+
+		rc = crt_rank_state_get(crt_group_lookup(NULL),
+				   doms[i].do_comp.co_rank, &state);
+		if (rc != 0) {
+			D_ERROR("failed to get swim for rank %u: %d\n",
+				doms[i].do_comp.co_rank, rc);
+			return rc;
+		}
+
+		if (state.sms_status == SWIM_MEMBER_DEAD ||
+		    state.sms_status == SWIM_MEMBER_INACTIVE) {
```

Comment:
Hmm, not sure if we should evict INACTIVE members. @dmiter, any suggestions, please?

Reply:
We should evict INACTIVE also IMHO. The INACTIVE state is possible now only during bootstrap process. So, this means the node was reboot and now it's in off or bootstrapping state. From my understanding this mean we should evict this node and reconnect it again with new configuration when it's go from INACTIVE state into ALIVE.

Reply:
Would this be a problem when we create a pool immediately after starting the server for the first time (and the servers have just printed "DAOS I/O server ... started on rank ...")? In that case, if the pool service leader's local swim hasn't pinged an INACTIVE member, and if that member is part of the pool, then the pool service will attempt to exclude this member immediately. Or no?

Reply:
Hmm, if this happens, then the "inactive" rank will be excluded, and later when it becomes active, it will be brought into the pool by reintegration. Anyway, IMHO, the leader should exclude any rank it can not reach during step up  to avoid extra logic and complexity.

Reply:
"If this happens", if we don't exclude the INACTIVE member, and if the PS leader attempts to invoke RPCs to the INACTIVE member, the RPCs shall reach the destination. No?

I think it'd be conceptually simpler if we only try to recover potentially lost SWIM events that ds_pool_crt_event_cb is interested in. That is:

    src == CRT_EVS_SWIM && type == CRT_EVT_DEAD && !pool_disable_evict

By the way, do we need to respect pool_disable_evict here in pool_svc_check_node_status?

For INACTIVE, I think we need a timeout mechanism to eventually generate (in swim) or simulate (in daos) a DEAD event, if the member fails to respond to SWIM messages. (The mechanism should preferably be in swim, I believe.)